### ISSUE TITLE
Observation mismatch when downloading subthreshold triggers

### ISSUE THREAD

**Issue by skverbrugge** [parsotat/BatAnalysis#37]
With the help of @dmopalmer and @sibasish1 ...
When running download_swift_trigger_data using a utc triggertime, the code searches for any observations with a start time within 300s (by default) of the triggertime parameter. However, this raises 2 possible issues: first, it could return the wrong observation with a closer start time and second, there are many observations that contain the same triggertime from many different observing runs over multiple orbits of the observatory. 

Source code with issue from batlib.py:

1848: `query = {'start_time': f"{tstart.isot}..{tend.isot}", 'fields': 'all'}`
1849: `nearest_obs_table = from_heasarc(**query)`
1850: `dt = Time(float(triggermjd), format="mjd") - Time(nearest_obs_table["START_TIME"], format="mjd")`
1851: `closest_obsid = nearest_obs_table["OBSID"][np.argmin(np.abs(dt))]`
1852: `if timewindow > np.abs(dt[np.argmin(np.abs(dt))].to("s")).value:`

For an example of the first problem, there are 2 BAT observations on 2022-10-10 from 01:27:31 - 01:46:57 (Obsid 01126853004) and from 01:47:00 - 02:00:27 (Obsid 00014981017). A trigger occuring at 01:39:00 would not return data from either observation because it is more than 300s away from either start time. However, a trigger occuring at 01:43:00 would return the data from 00014981017 because the start time is within 300s of the trigger time, but we are not interested in this data, rather we want the data from 01126853004 because it encloses the trigger time. 

For the second problem, we need to select the correct observation from the swiftmastr table similar to how `swinfo` returns a single Obs Sequence Number.

Here is a brief code snippet to demonstrate the first problem:
```python
import batanalysis as ba
import datetime

datadir = Path("/tmp/sb_test")
t0 = datetime.datetime(2022, 10, 10, 1, 39)
ba.datadir(“/tmp/sb_test", mkdir=True)
downloads = ba.download_swift_trigger_data(triggertime=t0)
print(downloads)
```
This code produces the following files:
<img width="214" height="336" alt="Image" src="https://github.com/user-attachments/assets/da9ac009-18c5-490d-8f83-d99db16d548d" />

Running swinfo for the times of interest produces this result:
<img width="338" height="194" alt="Image" src="https://github.com/user-attachments/assets/0514ca20-89c9-466a-873f-be9cce7ec0db" />

Obsids 01126883000 and 01126884000 are between 01:27:31 - 01:46:57 UTC, but the downloaded observation 00014981017 within is from the following observation from 01:47:00 - 02:00:27 UTC. This is because 01126884000 occurs at 01:43, which is within 300s of 01:47, so it takes that observation data instead of the real data with start time 01:27 (more than 300s away). Increasing the default triggerwindow would not change this behavior because it would simply encompass more start times and yet still return the closest one. Some consideration of the observation that actually encloses the triggertime needs to be included.

**Comment by parsotat** [comment]
@dmopalmer @skverbrugge @sibasish1 

I have just pushed a commit that updates the `from_heasarc` function to be compatible with astroquery 0.4.10 and constructs a simple TAP query to obtain the data from tables. It has the functionality to allow for a custom TAP query to be passed in which may be useful, particularly for correcting this bug here. 

In looking at the ADQL documentation (https://ivoa.net/documents/ADQL/20230418/PR-ADQL-2.1-20230418.html), there are ways to query both tables and correlate entries in one with respect to another which should allow us to look for failed triggers from the swifttdrss table, identify the start/stop time, and then find  the start/stop time of an actual obsid listed in the swiftmastr table  that encompasses the failed triggers' start/stop time.

**Comment by dmopalmer** [comment]
An observation corresponding to an OBSID is typically a number of 'snapshots' where the satellite is pointing in the same direction, interrupted by slews to other observations that are interleaved.  That is why yo have large-lying overlapping start_time-end_time intervals.

The attitude file should be the one that has the time of the trigger, since each observation has gaps in the attitude history when it is not the current target.  You need other pick the observation that was actually occurring at the time of the trigger.  (There may be more than one if the trigger occurred just before a slew.)

The gains and offsets are measured at slew time whenever they get sufficiently stale.  The definition of sufficiently stale was chosen before launch, 20+ years ago, and we never really checked the timescale on how fast these calibrations change.  There should be no problem with using a calibration for the observation instead of trying to find a nearer one from the next slew or the next time the gain or offset calibration is run.  (Typically, a later snapshot within the observation will have an updated gain or offset.)

**Comment by parsotat** [comment]
This means that the data provided in the swiftmastr table is not sufficient. Instead we will need to download the attitude file from each obsid with a start_time and end_time that overlap the sub threshold trigger time and then look at the GTIs of the files to determine which is actually applicable, correct @dmopalmer ?

As for the gain/offset, I am fine with using a relatively stale calibration file. It is also trivial (compared to the attitude file determination steps) to just download the gain/offset file that is closest in time to the sub threshold data.

**Comment by dmopalmer** [comment]
The `swiftmastr` table is insufficient.  However there is a more complete table maintained by the Swift project provided by the swifttools project.

``` python
import swifttools.swift_too as swto
import datetime
t0 = datetime.datetime(2022, 10, 10, 1, 39)

query = swto.ObsQuery(t0)
query.submit()
print(query.entries[0])
```
gives the individual snapshot that was occurring at that time.  (Or multiple snapshots possible if the ObsQuery is a time range).
```
+---------------------+---------------------+-------------+--------------------+--------------+--------------+
|          Begin Time |            End Time | Target Name | Observation Number | Exposure (s) | Slewtime (s) |
+---------------------+---------------------+-------------+--------------------+--------------+--------------+
| 2022-10-10 01:28:02 | 2022-10-10 01:47:28 |  GRB221009A |        01126853004 |         1070 |           96 |
+---------------------+---------------------+-------------+--------------------+--------------+--------------+
```

**Comment by parsotat** [comment]
please  test v2.1.0 of BatAnalysis which should correctly download the attitude file actually associated with a subthreshold TTE dataset and the nearest energy calibration file to the time when the TTE data was collected. 

Running this code chunk:
```
import batanalysis as ba
import datetime

datadir = Path("/tmp/sb_test")
t0 = datetime.datetime(2022, 10, 10, 1, 39)
ba.datadir(“/tmp/sb_test", mkdir=True)
downloads = ba.download_swift_trigger_data(triggertime=t0)
print(downloads)
```

produces:
```
{np.int32(1126883): [Swift_Data(username='anonymous',obsid='01126883000',quicklook='False',auxil='True',subthresh='True',itsdc='False'), Swift_Data(username='anonymous',obsid='00014981017',quicklook='False',auxil='False',bat='True'), Swift_Data(username='anonymous',obsid='01126853004',quicklook='False',auxil='True',bat='False')], 
np.int32(1126884): [Swift_Data(username='anonymous',obsid='01126884000',quicklook='False',auxil='True',subthresh='True',itsdc='False'), Swift_Data(username='anonymous',obsid='00014981017',quicklook='False',auxil='False',bat='True'), Swift_Data(username='anonymous',obsid='01126853004',quicklook='False',auxil='True',bat='False')]}
```

```
In [12]: !ls -l /tmp/sb_test/tdrss/2022_10/*
/tmp/sb_test/tdrss/2022_10/01126883000:
total 520
drwxr-xr-x  3 tparsota  wheel      96 Sep 29 14:51 00014981017
drwxr-xr-x  3 tparsota  wheel      96 Sep 29 14:51 01126853004
lrwxr-xr-x  1 tparsota  wheel      17 Sep 29 14:51 auxil -> 01126853004/auxil
lrwxr-xr-x  1 tparsota  wheel      15 Sep 29 14:51 bat -> 00014981017/bat
-rw-r--r--  1 tparsota  wheel    1564 Sep 29 14:51 sw01126883000msbal.fits.gz
-rw-r--r--  1 tparsota  wheel    2448 Sep 29 14:51 sw01126883000msbat.hk.gz
-rw-r--r--  1 tparsota  wheel  168789 Sep 29 14:51 sw01126883000msbevshsp_uf.evt.gz
-rw-r--r--  1 tparsota  wheel   63315 Sep 29 14:51 sw01126883000msbevtlsp.hk.gz
-rw-r--r--  1 tparsota  wheel    4213 Sep 29 14:51 sw01126883000msbevtssp.hk.gz
-rw-r--r--  1 tparsota  wheel    3525 Sep 29 14:51 sw01126883000msbhd.hk.gz
-rw-r--r--  1 tparsota  wheel    2163 Sep 29 14:51 sw01126883000msbno.fits.gz
-rw-r--r--  1 tparsota  wheel     892 Sep 29 14:51 sw01126883000mspob.cat.gz
drwxr-xr-x  3 tparsota  wheel      96 Sep 29 14:51 tdrss

/tmp/sb_test/tdrss/2022_10/01126884000:
total 496
drwxr-xr-x  3 tparsota  wheel      96 Sep 29 14:51 00014981017
drwxr-xr-x  3 tparsota  wheel      96 Sep 29 14:51 01126853004
lrwxr-xr-x  1 tparsota  wheel      17 Sep 29 14:51 auxil -> 01126853004/auxil
lrwxr-xr-x  1 tparsota  wheel      15 Sep 29 14:51 bat -> 00014981017/bat
-rw-r--r--  1 tparsota  wheel    1569 Sep 29 14:51 sw01126884000msbal.fits.gz
-rw-r--r--  1 tparsota  wheel    2455 Sep 29 14:51 sw01126884000msbat.hk.gz
-rw-r--r--  1 tparsota  wheel  161902 Sep 29 14:51 sw01126884000msbevshsp_uf.evt.gz
-rw-r--r--  1 tparsota  wheel   61339 Sep 29 14:51 sw01126884000msbevtlsp.hk.gz
-rw-r--r--  1 tparsota  wheel    4205 Sep 29 14:51 sw01126884000msbevtssp.hk.gz
-rw-r--r--  1 tparsota  wheel    3523 Sep 29 14:51 sw01126884000msbhd.hk.gz
-rw-r--r--  1 tparsota  wheel    2174 Sep 29 14:51 sw01126884000msbno.fits.gz
-rw-r--r--  1 tparsota  wheel     896 Sep 29 14:51 sw01126884000mspob.cat.gz
drwxr-xr-x  3 tparsota  wheel      96 Sep 29 14:51 tdrss
```

**Comment by parsotat** [comment]
@skverbrugge can this issue be closed due to the changes that were made in v2.1.0?

**Comment by skverbrugge** [comment]
Yes I think so
